src/levanter/infra/ray_tpu.py

```diff
@@ -0,0 +1,319 @@
+import dataclasses
+import logging
+import os
+import subprocess
+from dataclasses import dataclass
+from typing import Sequence
+
+import draccus
+import ray
+from ray.exceptions import NodeDiedError, RayError, RaySystemError, RayTaskError, WorkerCrashedError
+from ray.remote_function import RemoteFunction
+
+from levanter.infra.cli_helpers import make_docker_run_command
+
+
+# CF https://gist.github.com/allenwang28/e3400b9e9212b50aa1cda55ebeccea60
+
+logger = logging.getLogger("ray")
+
+
+@dataclass
+class _TpuInfo:
+    """Internal class to hold information about a TPU pod."""
+
+    name: str
+    state: str
+    kind: str
+
+
+# My kingdom for ADTs
```

Comment:
Ha! I've been using pydantic a bunch recently, and it's.. okay?

viz

class TpuRunResult(BaseModel):
   info: _TpuInfo

Gives you validation, JSON + schema support, etc. You can do tagged unions, but it feels weird with JSON in my experience. Here, just:

```
class TpuRunStatus(enum.StrEnum):
  success = 'success'
  preempted = 'preempted'
  failed = 'failed'

class TpuRunResult(BaseModel):
   status: TpuRunStatus
   info: _TpuInfo
   error: Optional[Exception]
```

Reply:
not too bad but TpuResult has the answer in it, so it's really more like Result<T, E> in rust...

Reply:
Ah yeah, you can kind of do it, but union support is kind of funky:

class TpuSuccess:
  type = Literal['success']
  result: Any

class TpuPreempted:
  type = Literal['preempted']

class TpuError:
  type = Literal['error']
  error: Exception

TpuResult = Annotated[
  Union[TpuSuccess, TpuPreempted, TpuError], Field(discrimator='type')]

Definitely a bit bulkier than the Rust version :p . It's a bit more reasonable when you think of the JSON serialization.

Reply:
(Of course this is all fine!)

Reply:
Yeah, truly Python is a beautiful language.